File: frontend/src/api/notification.ts
import { getWithRetry, putWithRetry, postWithRetry } from "./client";

export interface NotificationChannelConfigResponse {
  id: string;
  channel_type: "EMAIL" | "TELEGRAM" | "WEBHOOK";
  config: string; // JSON-serialized config
  enabled: boolean;
}

export interface NotificationDeliveryLogResponse {
  id: string;
  alert_id: string | null;
  channel: "EMAIL" | "TELEGRAM" | "WEBHOOK";
  recipient: string;
  title: string;
  message: string;
  priority: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
  status: "PENDING" | "SENT" | "FAILED";
  retry_count: number;
  max_retries: number;
  error_message: string | null;
  timestamp: string;
  sent_at: string | null;
}

export async function getChannels(): Promise<NotificationChannelConfigResponse[]> {
  return getWithRetry<NotificationChannelConfigResponse[]>("/api/v1/notifications/channels");
}

export async function updateChannel(
  channelType: string,
  payload: { config?: string; enabled?: boolean }
): Promise<NotificationChannelConfigResponse> {
  return putWithRetry<NotificationChannelConfigResponse>(
    `/api/v1/notifications/channels/${channelType}`,
    payload
  );
}

export async function triggerTestDispatch(
  channelType: string,
  payload: { channel: string; recipient: string; message: string }
): Promise<NotificationDeliveryLogResponse> {
  return postWithRetry<NotificationDeliveryLogResponse>(
    `/api/v1/notifications/channels/${channelType}/test`,
    payload
  );
}

export async function getDeliveryLogs(filters?: {
  channel?: string;
  status?: string;
  alert_id?: string;
  limit?: number;
}): Promise<NotificationDeliveryLogResponse[]> {
  let query = "";
  const params: string[] = [];
  if (filters?.channel && filters.channel !== "ALL") {
    params.push(`channel=${filters.channel}`);
  }
  if (filters?.status && filters.status !== "ALL") {
    params.push(`status=${filters.status}`);
  }
  if (filters?.alert_id) {
    params.push(`alert_id=${filters.alert_id}`);
  }
  if (filters?.limit) {
    params.push(`limit=${filters.limit}`);
  }
  if (params.length > 0) {
    query = `?${params.join("&")}`;
  }
  return getWithRetry<NotificationDeliveryLogResponse[]>(`/api/v1/notifications/logs${query}`);
}

File: backend/app/models/notification.py
from datetime import datetime
import uuid
from sqlalchemy import Column, String, Boolean, DateTime, Integer
from app.models.base import Base


class NotificationChannelConfig(Base):
    """
    SQLAlchemy model representing dynamic user/operator configurations for delivery channels.
    Includes state toggling and JSON-serialized credentials/endpoints.
    """

    __tablename__ = "notification_channel_configs"

    id = Column(
        String(50),
        primary_key=True,
        index=True,
        default=lambda: f"cfg-{uuid.uuid4().hex[:8]}",
    )
    channel_type = Column(
        String(20), unique=True, index=True, nullable=False
    )  # EMAIL, TELEGRAM, WEBHOOK
    config = Column(String(1000), nullable=False)  # JSON string representation
    enabled = Column(Boolean, default=True, nullable=False)


class NotificationDeliveryLog(Base):
    """
    SQLAlchemy model representing delivery logs, status tracking, and retry counts
    for multi-channel notifications.
    """

    __tablename__ = "notification_delivery_logs"

    id = Column(
        String(50),
        primary_key=True,
        index=True,
        default=lambda: f"log-{uuid.uuid4().hex[:8]}",
    )
    alert_id = Column(String(50), nullable=True, index=True)
    channel = Column(String(20), nullable=False, index=True)  # EMAIL, TELEGRAM, WEBHOOK
    recipient = Column(String(255), nullable=False)
    title = Column(String(255), nullable=False)
    message = Column(String(2000), nullable=False)
    priority = Column(
        String(20), nullable=False, index=True
    )  # LOW, MEDIUM, HIGH, CRITICAL
    status = Column(String(20), nullable=False, index=True)  # PENDING, SENT, FAILED
    retry_count = Column(Integer, nullable=False, default=0)
    max_retries = Column(Integer, nullable=False, default=3)
    error_message = Column(String(500), nullable=True)
    timestamp = Column(DateTime, default=datetime.utcnow, nullable=False, index=True)
    sent_at = Column(DateTime, nullable=True)
